### ISSUE TITLE
ivy-read `:preselect` should respect `ivy-case-fold-search-default`

### ISSUE THREAD

**Issue by Konubinix** [abo-abo/swiper#2770]
Ivy is great, and very well documented.

The documentation says exactly what `:preselect` does, so may be I am not looking in the right direction. 

First let's look at a common use case.

Using ivy, I want to complete against a collection

`(ivy-read "> " '("a" "bc" "b"))`

I can type "b" and see the results filtered on "bc" and "b", with "b" selected (because it is an exact match)

Let's say I activate the `ivy-case-fold-search-default`

I can run
`(ivy-read "> " '("A" "BC" "B"))`

And type "b" and have "B" selected (because it is an exact match, modulo the case).

Then, back to the first example, let's say I want to have the same behavior, but with "b" prefilled.

I then write:

`(ivy-read "> " '("a" "bc" "b") :preselect "b" :initial-input "b")`

And get the exact same result as if I did type "b" in the first example.

Then, based on that experience, I would like to naturally get the second example to work.

`(ivy-read "> " '("A" "BC" "B") :preselect "b" :initial-input "b")`

And I would expect to have "B" selected, while it is not the case.

I don't want to say that `:preselect` is bad documented or it should do something else than is it actually doing. I'm just saying that it looks natural, after playing with it a bit, to expect it to behave that way, meaning ignoring the case when ivy-case-fold-search-default is t.

What is your opinion? Maybe there is something I am missing here.

**Comment by basil-conto** [comment]
Are you sure this issue is about `:preselect` and not `:initial-input`?  [BTW, why are you specifying `:preselect` and `:initial-input` simultaneously?]

`:preselect` is documented as being matched first with `equal`, and then if that fails, with `string-match-p`, i.e. as a regexp.

So although I wouldn't mind making it smart-case-sensitive, that could be seen as a breaking change. Also, `:preselect` is a non-interactive programmatic feature, so it should be relatively easy for the caller to arrange for the right case to be used.

`:initial-input`, on the other hand, is an interactive feature - it should be as if the user had entered the corresponding text. So I think in its case it makes more sense to respect `ivy-case-fold-search-default`.

**Comment by Konubinix** [comment]
Hi,

Sorry, I realize I was unclear about what I want to achieve.

"Basil L. Contovounesios" <notifications@github.com> writes:

[...]

> `:initial-input`, on the other hand, is an interactive feature - it should be as if the user had entered the corresponding text. So I think in its case it makes more sense to respect `ivy-case-fold-search-default`.

This is exactly what I want :-): make ivy behave as if the user had
entered the corresponding text.

Now, the difference is:

1. If I run `(ivy-read "> " '("a" "bc" "b"))`, and press "b", then
   1. ivy will filter the collection to "bc" and "b"
   2. AND ivy will select "b", the exact match.

2. If I run `(ivy-read "> " '("a" "bc" "b") :initial-input "b")`, I
would expect to have the same behavior, but:
   1. ivy will indeed filter the collection to "bc" and "b", like
   expected
   2. ivy will select "bc", which is not the exact match.


I got used to add `:preselect "b"` to artificially get the pre-selection
of "b", and this work around does the trick, except when the case is
mixed (hence the title of the issue).

But indeed, if :initial-input is already meant to behave as if the user
had entered the corresponding text, then no need to bother with
:preselect.

Is it clearer what I want to achieve and the trouble I have to get to
it?

My best
-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

**Comment by basil-conto** [comment]
> Is it clearer what I want to achieve and the trouble I have to get to it?

Yes, and I would also expect this behaviour. But Ivy's indexing, preselection, and initial input code is a tangled mess in my eyes, so I'm afraid I can't help much with this. :)

**Comment by Konubinix** [comment]
"Basil L. Contovounesios" <notifications@github.com> writes:

>> Is it clearer what I want to achieve and the trouble I have to get to it?

> Yes, and I would also expect this behaviour. But Ivy's indexing, preselection, and initial input code is a tangled mess in my eyes, so I'm afraid I can't help much with this. :)

Ok, thanks for helping me clarify the issue anyway :-)

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

**Comment by Konubinix** [comment]
I rephrased the title to make clearer what I would love ivy to do.
